std/array.d

```diff
@@ -342,6 +342,110 @@ then the result will contain the value of the last pair for that key in r.
 See_Also: $(XREF typecons, Tuple)
  */
 
+/**
+ * Initialize a static array of the given size from the range $(D r).
+ *
+ * If `r` has more than `n` elements, only the first `n` are copied.
+ * If `r` has less than `n` elements, the remaining spaces in the static array
+ * will have default values.
+ *
+ * Params:
+ *      r = range (or aggregate with $(D opApply) function) whose elements are
+ *      copied into the static array.
+ *
+ * Returns:
+ *      static array of size `n`
+ */
+ForeachType!Range[n] array(size_t n, Range)(Range r)
+if (isIterable!Range)
+{
+    typeof(return) result;
+
+    // manually count our index instead of using foreach(i, e; r) because:
+    // 1. for a range, we would need to use enumerate
+    // 2. for opApply with no indexed overload, we would need to count anyways
+    size_t i;
+    foreach (e; r)
+    {
+        if (i >= n) break; // array is full, ignore remaining elements
+        result[i++] = e;
+    }
+
+    return result;
+}
+
+///
+@safe pure nothrow unittest
+{
+    import std.range : only;
+    int[3] a = only(1, 2, 3).array!3;
+    assert(a == [ 1, 2, 3 ]);
+}
+
+/// `array!n` only takes `n` elements, even if the range has (infinitely) more.
+@safe pure nothrow unittest
+{
+    import std.range : repeat;
+    int[4] a = repeat(5).array!4;
+    assert(a == [ 5, 5, 5, 5 ]);
+}
+
+/// `array!n` leaves remaining elements default if the range is smaller than `n`
+@safe pure nothrow unittest
+{
+    import std.range : only;
+    int[5] a = only(1,2,3).array!5;
+    assert(a == [ 1, 2, 3, 0, 0 ]);
+}
+
+/// `array!n` works with `opApply` as well as a range interface
+unittest
+{
+    static struct OpApply
+    {
+        int opApply(int delegate(ref int) dg)
+        {
+            int res;
+            foreach(i; 0..10)
+            {
+                res = dg(i);
+                if(res) break;
+            }
+
+            return res;
+        }
+    }
+
+    int[5] a = OpApply().array!5;
+    assert(a == [ 0, 1, 2, 3, 4 ]);
+}
+
+/**
+ * Initialize a static array from the given elements (essentially a static
+ * array literal).
+ *
+ * Params:
+ *      values = elements of the static array
+ *
+ * Returns: static array of size `T.length`
+ */
+auto staticArray(T...)(T values)
```

Comment:
Doh. Yeah, that's much simpler. Thanks!

On March 16, 2016 11:08:51 PM EDT, JakobOvrum <notifications@github.com> wrote:
>> +    }
>> +
>> +    int[5] a = OpApply().array!5;
>> +    assert(a == [ 0, 1, 2, 3, 4 ]);
>> +}
>> +
>> +/**
>> + * Initialize a static array from the given elements (essentially a
>static
>> + * array literal).
>> + *
>> + * Params:
>> + *      values = elements of the static array
>> + *
>> + * Returns: static array of size `T.length`
>> + */
>> +auto staticArray(T...)(T values)
>
>```d
>CommonType!T[T.length] staticArray(T...)(T values) if
>(is(CommonType!T))
>{
>    typeof(return) result = [values];
>    return result;
>}
>```
>
>---
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/D-Programming-Language/phobos/pull/4090/files/67c6a7b1ae5a0f41c43c45f80048a19be469e77f#r56450014

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.